fastjet/clustersequence.go

```diff
@@ -69,8 +72,63 @@ func NewClusterSequence(jets []Jet, def JetDefinition) (*ClusterSequence, error)
 	return cs, err
 }
 
+//returns the number of exclusive jets that would have been obtained
+//running the algorithm in exclusive mode with the given dcut
+func (cs *ClusterSequence) NrExclusiveJets(dcut float64) int {
+	//first locate where clustering would have stopped
+	i := len(cs.history) - 1 //last jet
+	for ; 0 <= i; i-- {
+		if cs.history[i].maxdij <= dcut {
+			break
+		}
+	}
+	stoppt := i + 1
+	njets := 2*cs.initn - stoppt
+	return njets
+}
+
 func (cs *ClusterSequence) ExclusiveJets(dcut float64) ([]Jet, error) {
-	panic("not implemented")
+	njets := cs.NrExclusiveJets(dcut)
+	return cs.ExclusiveJetsUpTo(njets)
+}
+
+func (cs *ClusterSequence) ExclusiveJetsUpTo(njets int) ([]Jet, error) {
+	var err error
+	if njets > cs.initn {
+		err = errors.New("Requested " + strconv.Itoa(njets) + " exclusive jets, but there are only " + strconv.Itoa(cs.initn) + " particles in the event")
+		panic(err)
+	}
+	//calculate the point where we have to stop the clustering
+	//relation between ptstop, njets assumes one extra jet disappears
+	//at each clustering
+	stoppt := 2*cs.initn - njets
+	//make sure it's safe when more jets are requested than there are particles
+	if stoppt < cs.initn {
+		stoppt = cs.initn
+	}
+	//additional checking
+	if 2*cs.initn != len(cs.history) {
+		err = errors.New("2*initial nr jets != size of history -- this endagers internal assumptions!")
+		panic(err)
+	}
+
+	//now go forwards and reconstitute the jets that we have --
+	//basically for any history element, see if the parent jets to
+	//which it refers were created before the stopping point -- if they
+	//were then add them to the list, otherwise they are subsequent
+	//recombinations of the jets that we are looking for
+	ljets := make([]Jet, 0)
```

Comment:
yes, if a good guess for the capacity can be provided, then we should file it.
that's that much less re-allocations to perform at runtime.